src/solver/opamBuiltin0install.ml

```diff
@@ -90,25 +91,53 @@ let reconstruct_universe universe selections =
 type options = {
   drop_installed_packages : bool;
   prefer_oldest : bool;
+  handle_avoid_version : bool;
+  prefer_installed : bool;
 }
 
 let parse_criteria criteria =
-  let default = {drop_installed_packages = false; prefer_oldest = false} in
-  match criteria with
-  | "" -> default
-  | "+removed" -> {drop_installed_packages = true; prefer_oldest = false}
-  | "+count[version-lag,solution]" -> {drop_installed_packages = false; prefer_oldest = true}
-  | "+removed,+count[version-lag,solution]" ->
-    {drop_installed_packages = true; prefer_oldest = true}
-  | _ ->
-    OpamConsole.warning "Criteria '%s' is not supported by the 0install solver" criteria;
-    default
+  let default =
+    {
+      drop_installed_packages = false;
+      prefer_oldest = false;
+      handle_avoid_version = false;
+      prefer_installed = false;
+    }
+  in
+  let rec parse default = function
```

Comment:
`OpamBuiltinZ3` contains a `Syntax` module that parses criteria. We can move it into `OpamCudfSolverSig` and use it directly instead of having a parsing function per solver.

Reply:
done. I don't think this should be in `OpamCudfSolverSig` as that module is meant for signatures not implementations but it fits really well into a new `OpamCudfCriteria` module